src/statsmodels/formula.jl

```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
 end
 
 function ModelMatrix(mf::ModelFrame)
-    trms = mf.terms
-    aa = Any[Any[ones(size(mf.df,1), @compat(Int(trms.intercept)))]]
-    asgn = zeros(Int, @compat(Int(trms.intercept)))
-    fetrms = Bool[isfe(t) for t in trms.terms]
-    if trms.response unshift!(fetrms, false) end
-    ff = trms.factors[:, fetrms]
-    ## need to be cautious here to avoid evaluating cols for a factor with many levels
-    ## if the factor doesn't occur in the fetrms
-    rows = Bool[x != 0 for x in sum(ff, 2)]
-    ff = ff[rows, :]
-    cc = [cols(col) for col in columns(mf.df[:, rows])]
-    for j in 1:size(ff,2)
-        trm = cc[round(Bool, ff[:, j])]
-        push!(aa, trm)
-        asgn = vcat(asgn, fill(j, nc(trm)))
+    ## TODO: this method makes multiple copies of the data in the ModelFrame:
+    ## first in term_cols (1-2x per evaluation term, depending on redundancy),
+    ## second in constructing the matrix itself.
+    
+    ## Map eval. term name + redundancy bool to cached model matrix columns
+    eterm_cols = Dict{Tuple{Symbol,Bool}, Array{Float64}}()
+    ## Accumulator for each term's vector of eval. term columns.
+    mm_cols = Vector{Array{Float64}}[]
+    mf.terms.intercept && push!(mm_cols, Matrix{Float64}[ones(Float64, size(mf.df,1), 1)])
+    factors = round(Bool, mf.terms.factors)
+
+    ## turn each term into a vector of mm columns for its eval. terms, using
+    ## "promoted" full-rank versions of categorical columns for non-redundant
+    ## eval. terms:
+    for (i_term, term) in enumerate(mf.terms.terms)
+        ## Skip non-fixed-effect terms
+        isfe(term) || continue
+        term_cols = Array{Float64}[]
+        ## adjust term index if there's a response in the formula (mf.terms.terms
+        ## lists only non-response terms, but mf.terms.factors and .non_reundant_terms
+        ## has first column for the response if it's present)
+        i_term += mf.terms.response
+        ## Pull out the eval terms, and the non-redundancy flags for this term
+        eterms = mf.terms.eterms[factors[:, i_term]]
+        non_redundant = mf.non_redundant_terms[factors[:, i_term], i_term]
+        ## Get cols for each eval term (either previously generated, or generating
+        ## and storing as necessary)
+        for et_and_nr in zip(eterms, non_redundant)
+            haskey(eterm_cols, et_and_nr) || 
+                setindex!(eterm_cols,
+                          modelmat_cols(et_and_nr[1], mf, non_redundant=et_and_nr[2]),
+                          et_and_nr)
+            push!(term_cols, eterm_cols[et_and_nr])
+        end
+        push!(mm_cols, term_cols)
     end
-    ModelMatrix{Float64}(hcat([expandcols(t) for t in aa]...), asgn)
+
+    ## TODO: could this be made more efficient by
+    ## first computing mm_col_term_nums, initializing mm, and directly indexing?
+    mm = hcat([expandcols(tc) for tc in mm_cols]...)
```

Comment:
Well, it's new code AFAICT, so it's relevant to this PR. What I meant is that splatting will force compiling `hcat` for each type/number combination of its argument. Since these vary a lot depending on the data, this will trigger the compilation of many functions. See https://github.com/JuliaLang/julia/issues/13722

So indeed the TODO is right that it would be better to create the matrix first, and then fill it with a loop. Not sure whether there's a simpler solution.